### ISSUE TITLE
Use modules and require_relative

### ISSUE THREAD

**Issue by amiika** [dolphinOfDelphi/20210303-polyphony#7]
Hi! Awesome work. Looking forward to learn a lot from you :)

As a friendly suggestion I think you could use require_relative, modules and include instead of eval_file in your project. 

Check out Petal extension for Sonic Pi (https://github.com/siaflab/petal). Petal has a really nice project structure and i personally learned a lot from it. 

I used similar project structure in my project Ziffers (https://github.com/amiika/ziffers), but its more hacky and not very rubular.

**Comment by d0lfyn** [comment]
Hi! Thanks for taking a look at my project, and woah, I have lots to learn from _you_! \:D

I've mainly stuck to what's included in the Sonic Pi documentation. I understand that the proposed changes would create dedicated namespaces, which could help organise things, but isn't guaranteed to work in future releases of Sonic Pi (you may have seen [this discussion](https://in-thread.sonic-pi.net/t/why-define-macro/533/2?u=d0lfyn) before). Things may have been different several years ago, but Sonic Pi has stayed its course as a non-object-oriented environment.

Thanks for linking me to Petal and to Ziffers. They look like interesting projects, whose project structures look to be more object-oriented.

**Comment by amiika** [comment]
Yes. I went astray to the ruby dark side years ago :D You can always monkey patch things in Ruby if you don't like it. And yeah ... i'm aware that things might break in the future, but for me it's really a presumption. At least in Sonic Pi you can always use the older version.

Modules are not itself object oriented, more like containers that can hold methods and share scoped variables. I don't see how those could break in the future, but those are indeed out of Sonic Pi's documentation.

**Comment by d0lfyn** [comment]
Ahaha, I see. I'm still new to Ruby and to Sonic Pi. I'd like to experiment with modules for grouping methods exclusively (whereas my global variables all go into one file for ease of access). I can see that modules would work for enclosing calculation methods that don't make any reference to the Sonic Pi library, but that limits my use for them.

**Comment by amiika** [comment]
Sonic Pi methods work just fine in Modules. In the end Sonic Pi methods are just ruby methods. Think of it like there is two separate things: Ruby interpreter and Sonic Pi runtime. 

If you want to introduce new methods that runs in Sonic Pi runtime then you have to use load / require / require_relative to introduce things in to the runtime. 

So if you want to create background libraries create a new Module in a new file like:

```
module MyModule
  module Test
    def foo # Use def in modules since define is a Sonic Pi runtime thing
      sample :ambi_drone
    end
  end
end

# Include in the same file or other file but never in Sonic Pi.
include MyModule::Test 
```

and then introduce it to the Sonic Pi like this:

```
load "~/ziffers/moduletest.rb" # Use load to make like changes to moduletest.rb
# use require or require_relative for finished modules

foo #works
```
This was a silly example, best use case for external modules is of course background operations needed to make the composition in Sonic Pi, like things related to music theory or transformations etc.

**Comment by amiika** [comment]
Here's a more useful example, which also shows the _unlimited power_ of the ruby language:

```
module NoteHelper
  include SonicPi::Lang::WesternTheory
  def to_h
    (note_info self).to_h
  end
end

Integer.include NoteHelper
```
This is not only then extending Sonic Pi, but also Ruby primitive itself. After you require this in Sonic pi you have this naughty little helper that makes things fast and fun:
```
require "~/notehelper.rb"
print 70.to_h
```

**Comment by d0lfyn** [comment]
Amazing, so that's how things are connected! Alright, I think I'll do a complete refactoring of my code for v0.4.0. I'll try to figure it out from here. Thanks for taking the time to point out these aspects of the language I was never aware of!

I'm excited that my code will finally show up in VSCode as navigable Ruby modules and methods, among other things.

Also, I'd very much like to translate my musical building blocks (i.e. notes and motifs) into instantiable classes like in `petal_data.rb` from Petal. The only problem is that to my knowledge, Sonic Pi's time state can't store instances. Is there another way to approach this problem?

**Comment by amiika** [comment]
Classes and instances work fine in separate modules - but not in Sonic Pi runtime. Petal is using classes and it runs just fine, cos you don't have to call those classes and instances directly. Instead you are using methods that do things or return other than class instances.

So if you want to use classes in background library it works as long as you don't try to return the instance to Sonic Pi runtime. When calling some method from Sonic Pi return the data as hash objects, arrays, etc. For example in Ziffers im using array of hash objects as a data structure. I didnt use classes but you could transform class instances to hash objects and return those.

**Comment by amiika** [comment]
Just tried to hack some generic instance to hash method:

```
module ClassTest

class DataClass
  def initialize(first,second)
    @first = first
    @second = second
  end
  def to_h
    instance_variables.map { |v| [v[1..-1].to_sym, instance_variable_get(v)] }.to_h
  end
end

def getMyStuff
  data = DataClass.new("foo","bar")
  data.to_h
end

end

include ClassTest

```

Works. But then also realized that [Struct](https://ruby-doc.org/core-3.0.0/Struct.html) actually implements .to_h out of the box.

```
Customer = Struct.new(:name, :address) do
  def greeting
    "Hello #{name}!"
  end
end

dave = Customer.new("Dave", "123 Main")
print dave.name     #=> "Dave"
print dave.greeting #=> "Hello Dave!"
print dave.to_h
```

And to my suprise this also works in Sonic Pi in runtime! ... but kind of makes sense since its not actually Class class, but a subclass of Struct.

**Comment by d0lfyn** [comment]
Some considerations lead me to think that though it would be nice to have instance methods, I should stick with my current implementation, because:

- frequently converting instances to hashes and back to instances when I want to use their methods (which is often) would incur a performance penalty
- Struct comes with a modifier method, but my hashes are always frozen and final

Thanks for sharing your knowledge. It's best to make an informed decision, knowing the alternatives.

**Comment by amiika** [comment]
I think you can actually make those classes in the background modules and instantiate them from the Sonic Pi. That way you could also use instance methods. It's 'against' Sonic Pi's spirit, but hey things are meant to be bent and broken :)

**Comment by d0lfyn** [comment]
I'll give it a try! \:) This re-implementation won't make a difference to the music, which is always my focus, so I'll put it in an experimental branch.

**Comment by d0lfyn** [comment]
Update: I'm re-factoring the entire project to use modules and classes for v0.4.0. This will take at least a few days, as I'm also taking the opportunity to document and re-think everything. @amiika I'll mention you and Ziffers (as well as Petal) in my change log for your suggestion. Thanks again!

**Comment by d0lfyn** [comment]
The codebase has been refactored to use modules and _load_, as opposed to require_relative. The justification for this latter decision is that loading files instead of requiring them forces a reload of all modules on every run. This means that the program will immediately reflect changes to the source code, without requiring that the Sonic Pi environment be rebooted.

I have taken the time to document the code with YARD, identifying variable types and elaborating on the more complex functions.

The readme and changelog have been updated. I made sure to mention @amiika for his suggestions and explanations, without which I wouldn't have considered this rewrite. I've learnt lots and made the code more maintainable and also navigable on GitHub.

The Tests module is a stub. Future iterations should address testing, but it is not my priority at the moment. Though the music sounds fine, it is important that the functions that produce it are consistent and behave as expected. Therefore, putting off testing is not recommendable.

This issue will be resolved with the merging of v0.4.0 into the main branch.
